client/stunlistener/listener_linux.go

```diff
@@ -0,0 +1,366 @@
+				continue
+			}
+			msg := &stun.Message{
+				Raw: buf,
+			}
+			err = msg.Decode()
+			if err != nil {
+				log.Errorf("listenerToMux got an error while parsing stun message: %s", err)
+				continue
+			}
+
+			err = mux.HandleSTUNMessage(msg, a)
+			if err != nil {
+				log.Errorf("listenerToMux got an error while handling stun message: %s", err)
+			}
+		}
+	}
+}
+
+// updateRouter updates the listener routing table client
+// this is needed to avoid outdated information across different client networks
+func (s *StunListener) updateRouter() {
+	ticker := time.NewTicker(15 * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case <-ticker.C:
+			router, err := netroute.New()
+			if err != nil {
+				log.Errorf("failed to create and update packet router for stunListener: %s", err)
+				continue
+			}
+			s.routerMux.Lock()
+			s.router = router
+			s.routerMux.Unlock()
+		}
+	}
+}
+
+// LocalAddr returns an IPv4 address using the supplied port
+func (s *StunListener) LocalAddr() net.Addr {
+	// todo check impact on ipv6 discovery
+	return &net.UDPAddr{
+		IP:   net.IPv4zero,
+		Port: s.port,
+	}
+}
+
+// SetDeadline sets both the read and write deadlines associated with the ipv4 and ipv6 Conn sockets
+func (s *StunListener) SetDeadline(t time.Time) error {
+	err := s.conn4.SetDeadline(t)
+	if err != nil {
+		return fmt.Errorf("s.conn4.SetDeadline error: %s", err)
+	}
+	if s.conn6 == nil {
+		return nil
+	}
+
+	err = s.conn6.SetDeadline(t)
+	if err != nil {
+		return fmt.Errorf("s.conn6.SetDeadline error: %s", err)
+	}
+	return nil
+}
+
+// SetReadDeadline sets the read deadline associated with the ipv4 and ipv6 Conn sockets
+func (s *StunListener) SetReadDeadline(t time.Time) error {
+	err := s.conn4.SetReadDeadline(t)
+	if err != nil {
+		return fmt.Errorf("s.conn4.SetReadDeadline error: %s", err)
+	}
+	if s.conn6 == nil {
+		return nil
+	}
+
+	err = s.conn6.SetReadDeadline(t)
+	if err != nil {
+		return fmt.Errorf("s.conn6.SetReadDeadline error: %s", err)
+	}
+	return nil
+}
+
+// SetWriteDeadline sets the write deadline associated with the ipv4 and ipv6 Conn sockets
+func (s *StunListener) SetWriteDeadline(t time.Time) error {
+	err := s.conn4.SetWriteDeadline(t)
+	if err != nil {
+		return fmt.Errorf("s.conn4.SetWriteDeadline error: %s", err)
+	}
+	if s.conn6 == nil {
+		return nil
+	}
+
+	err = s.conn6.SetWriteDeadline(t)
+	if err != nil {
+		return fmt.Errorf("s.conn6.SetWriteDeadline error: %s", err)
+	}
+	return nil
+}
+
+// Close closes the underlying ipv4 and ipv6 conn sockets
+func (s *StunListener) Close() error {
+	errGrp := errgroup.Group{}
+	errGrp.Go(s.conn4.Close)
+	if s.conn6 != nil {
+		errGrp.Go(s.conn6.Close)
+	}
+	return errGrp.Wait()
+}
+
+// read start a read loop for a specific receiver and sends the packet to the packetDemux channel
+func (s *StunListener) read(receiver receiver) {
+	for {
+		buf := make([]byte, 1500)
```

Comment:
Is this loop can cause performance issue? If yes we should use the sync.Pool instead of create new slice all the time.